### ISSUE TITLE
SPI: one of the half-duplex modes is not supported

### ISSUE THREAD

**Issue by bugadani** [esp-rs/esp-hal#2687]
I can't actually tell whether the half-duplex mode we support is 3 or 4-wire single-bit. Our data modes are defined as:

```rust
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
#[cfg_attr(feature = "defmt", derive(defmt::Format))]
pub enum SpiDataMode {
    /// `Single` Data Mode - 1 bit, 2 wires.
    Single,
    /// `Dual` Data Mode - 2 bit, 2 wires
    Dual,
    /// `Quad` Data Mode - 4 bit, 4 wires
    Quad,
    #[cfg(spi_octal)]
    /// `Octal` Data Mode - 8 bit, 8 wires
    Octal,
}
```

This suggests 4-wire half-duplex mode (MOSI, MISO both transfer data). This is essentially full-duplex with state machine control.

The ESP32 TRM explicitly calls out a different single-bit half-duplex mode:

> The three-line half-duplex communication differs from four-line half-duplex communication in that the reception and transmission shares one signal bus

The other TRMs call out three-line half-duplex mode in a table, e.g. esp32-c6 v1.0 table 28-4.

Additionally, the hardware supports command/address/dummy phases in full-duplex transfer.

---

~~This is actually more general than the hardware supports (i.e. it's possible to generate invalid configuration). The set of options is roughly:~~

```rust
enum TransferMode {
    ThreeWireHalfDuplexSpi,
    FourWireHalfDuplexSpi,
    FullDuplexSpi,
    Qpi, // 4 bit command, address, data
    Opi, // 8 bit command, address, data
    WideSpiNarrowAddress(2/4/8 data bits), // 1 bit command, 1 bit address
    WideSpiWideAddress(2/4/8 data bits),   // 1 bit command, N bit address
}
```

**Comment by bugadani** [comment]
That being said, the S2 TRM explicitly says as an example, that this is supported:

![image](https://github.com/user-attachments/assets/035885a2-8eb6-4dfc-be3a-2206a993e2d2)

😩

**Comment by bjoernQ** [comment]
I think this boils down to just have a way to enable SIO

Easiest might be to just add a config option for that

**Comment by bugadani** [comment]
What would be the effect of that? Every half-duplex transfer would now be 3-wire, regardless of DataMode?

**Comment by bjoernQ** [comment]
Yes configuring it per transfer was the other option I thought about.

We shouldn't allow mixing any other DataMode with 3-wire mode in one transfer, then. Since those functions are fallible already, we can probably just check it and return a new error variant if needed

**Comment by bjoernQ** [comment]
I guess we reached consent here? (i.e. add a data mode for three-wire SPI)

If yes, I can assign this to me and remove the "investigation" label
